### PR TITLE
add API for getting link transforms

### PR DESCRIPTION
**Pull request by PeterMitrano**
not sure why there are formatting changes? pretty sure the clangformat file is being used?

### CODE REVIEW

File: jacobian_follower/include/jacobian_follower/jacobian_follower.hpp [hunk 1]
```diff
@@ -171,11 +170,30 @@ class JacobianFollower {
...
+  collision_detection::CollisionResult checkCollision(planning_scene::PlanningScenePtr planning_scene,
+                                                      robot_state::RobotState const &state);
 
   bool check_collision(moveit_msgs::PlanningScene const &scene_msg, moveit_msgs::RobotState const &start_state);
 
   PointSequence get_tool_positions(std::vector<std::string> tool_names, moveit_msgs::RobotState const &state_msg);
 
   void debugLogState(std::string prefix, robot_state::RobotState const &state);
+
+  Eigen::Matrix4Xd getLinkToRobotTransform(std::vector<std::string> joint_names, std::vector<double> joint_positions,
```
Comment (bsaund) [inline jacobian_follower/include/jacobian_follower/jacobian_follower.hpp:182]: arguments can be `const &`

File: jacobian_follower/src/jacobian_follower/jacobian_follower.cpp [hunk 2]
```diff
@@ -812,3 +813,46 @@ PointSequence JacobianFollower::get_tool_positions(std::vector<std::string> tool
...
+}
+
+std::vector<std::vector<Eigen::Matrix4Xd>> JacobianFollower::batchGetLinkToRobotTransforms(
+    std::vector<std::vector<std::string>> const &joint_names, std::vector<std::vector<double>> const &joint_positions,
+    std::vector<std::string> const &link_names) const {
+  auto state = std::make_shared<robot_state::RobotState>(model_);
+  std::vector<std::vector<Eigen::Matrix4Xd>> transforms;
+  auto const batch_size = joint_names.size();
+  for (auto b{0u}; b < batch_size; ++b) {
+    auto const joint_names_b = joint_names[b];
```
Comment (bsaund) [inline jacobian_follower/src/jacobian_follower/jacobian_follower.cpp:851]: 1) Remove these intermediate allocations. `transforms.emplace_back(getLinkToRobotTransforms(joint_names[b], joint_positions[b], state, link_names));` reads just as cleanly, and doesn't require your two lines `auto const joint_names_b = joint_names[b];`
2) If you really want these intermediate variables for some reason, they can be references `auto const &joint_names_b = joint_names[b];`

File: jacobian_follower/src/jacobian_follower/jacobian_follower.cpp [hunk 3]
```diff
@@ -812,3 +813,46 @@ PointSequence JacobianFollower::get_tool_positions(std::vector<std::string> tool
...
+std::vector<Eigen::Matrix4Xd> JacobianFollower::getLinkToRobotTransforms(
+    std::vector<std::string> const &joint_names, std::vector<double> const &joint_positions,
+    std::vector<std::string> const &link_names) const {
+  auto state = std::make_shared<robot_state::RobotState>(model_);
+  return getLinkToRobotTransforms(joint_names, joint_positions, state, link_names);
+}
+
+std::vector<std::vector<Eigen::Matrix4Xd>> JacobianFollower::batchGetLinkToRobotTransforms(
+    std::vector<std::vector<std::string>> const &joint_names, std::vector<std::vector<double>> const &joint_positions,
+    std::vector<std::string> const &link_names) const {
```
Comment (bsaund) [inline jacobian_follower/src/jacobian_follower/jacobian_follower.cpp:846]: Add some input validation: 
```if joint_names.size() != joint_positions.size() {throw something}```

If you throw inside a pybind function, what does the user see? Hopefully some useful error message.

File: jacobian_follower/src/jacobian_follower/jacobian_follower.cpp [hunk 4]
```diff
@@ -812,3 +813,46 @@ PointSequence JacobianFollower::get_tool_positions(std::vector<std::string> tool
...
+  robot_state::RobotState state(model_);
+  state.setVariablePositions(joint_names, joint_positions);
+  auto const &transform = state.getGlobalLinkTransform(link_name);
+  return transform.matrix();
+}
+
+std::vector<Eigen::Matrix4Xd> JacobianFollower::getLinkToRobotTransforms(
+    std::vector<std::string> const &joint_names, std::vector<double> const &joint_positions,
+    robot_state::RobotStatePtr state, std::vector<std::string> const &link_names) const {
+  state->setVariablePositions(joint_names, joint_positions);
```
Comment (bsaund) [inline jacobian_follower/src/jacobian_follower/jacobian_follower.cpp:827]: Does this already throw a useful error on invalid input, visible through pybind? If not, let's figure out a way to give the user a hint

**Review by bsaund — COMMENTED** [review]
Minor changes

File: jacobian_follower/src/jacobian_follower/jacobian_follower.cpp (hunk 2 above)
Comment (PeterMitrano) [reply jacobian_follower/src/jacobian_follower/jacobian_follower.cpp:851]: I'd be $100 the compiler removes the allocation & copy
but still I agree with the suggested change

File: jacobian_follower/src/jacobian_follower/jacobian_follower.cpp (hunk 3 above)
Comment (PeterMitrano) [reply jacobian_follower/src/jacobian_follower/jacobian_follower.cpp:846]: https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html?highlight=exception

even better, you can catch them as python exceptions!

Comment (bsaund) [reply jacobian_follower/src/jacobian_follower/jacobian_follower.cpp:848]: Amazing

File: jacobian_follower/include/jacobian_follower/jacobian_follower.hpp [hunk 5]
```diff
@@ -196,4 +196,14 @@ class JacobianFollower {
       std::vector<std::string> const &link_names) const;
 
   std::vector<std::string> getLinkNames() const;
+
+  template <typename A, typename B>
+  void validateNamesAndPositions(const std::vector<A> &joint_names, const std::vector<B> &joint_positions) const {
```
Comment (bsaund) [inline jacobian_follower/include/jacobian_follower/jacobian_follower.hpp:201]: A bit misleading error message, as in one case this checks the length of joint_names, and in another this checks the batch dimension. 

Still probably enough to point the user to debug the correct variables though

**Review by bsaund — APPROVED** [review]

**Pull request merged**